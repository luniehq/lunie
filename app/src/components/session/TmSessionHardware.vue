<template>
  <SessionFrame :icon="`vpn_key`">
    <h2 class="session-title">Use my Ledger Nano</h2>

    <template v-if="session.browserWithLedgerSupport">
      <div class="session-main">
        <HardwareState :loading="status === `connect` ? false : true">
          <template v-if="isWindows && !hasHIDEnabled">
            Due to recent Ledger updates, using a Ledger on Windows now requires
            "Experimental Web Platform features" to be enabled.
            <template v-if="isChrome">
              <br />
              <br />
              <p>
                Please copy the link below into a new tab and set the
                "Experimental Web Platform features" flag to "Enabled":
              </p>
              <div
                v-clipboard:copy="hidFeatureLink"
                v-clipboard:success="() => onCopy()"
                class="copy-feature-link"
              >
                {{ hidFeatureLink }}
                <i
                  class="material-icons notranslate copied"
                  :class="{ active: copySuccess }"
                >
                  check
                </i>
              </div>
              <br />
              <br />
            </template>
          </template>
          <template v-else-if="isLinux">
            Since we switched to WebUSB Linux users may experience connection
            issues with their devices.
            <br />
            <br />
            Please visit the following site to learn more about how to fix them:
            <div
              v-clipboard:copy="linuxLedgerConnectionLink"
              v-clipboard:success="() => onCopy()"
              class="copy-feature-link"
            >
              {{ linuxLedgerConnectionLink }}
              <i
                class="material-icons notranslate copied"
                :class="{ active: copySuccess }"
              >
                check
              </i>
            </div>
            <br />
            <br />
          </template>
          <template v-else-if="status === `connect` || status === `detect`">
            <p>
              Please plug in your Ledger&nbsp;Nano and open the Cosmos Ledger
              app
            </p>
          </template>
          <p v-if="connectionError" class="error-message">
            {{ connectionError }}
          </p>
          <TmBtn
            :value="submitCaption"
            :disabled="
              status === `connect` || (isWindows && !hasHIDEnabled)
                ? false
                : `disabled`
            "
            @click.native="signIn()"
          />
        </HardwareState>
      </div>
    </template>

    <div v-else class="session-main">
      <p>
        Please use Chrome or Brave. Ledger is not supported in this browser.
      </p>
    </div>
  </SessionFrame>
</template>

<script>
import TmBtn from "common/TmBtn"
import { mapState, mapGetters } from "vuex"
import HardwareState from "common/TmHardwareState"
import SessionFrame from "common/SessionFrame"
import { getAddressFromLedger } from "scripts/ledger"
import * as Sentry from "@sentry/browser"

export default {
  name: `session-hardware`,
  components: {
    TmBtn,
    SessionFrame,
    HardwareState,
  },
  data: () => ({
    status: `connect`,
    connectionError: null,
    address: null,
    copySuccess: false,
    hidFeatureLink: `chrome://flags/#enable-experimental-web-platform-features`,
    linuxLedgerConnectionLink: `https://support.ledger.com/hc/en-us/articles/360019301813-Fix-USB-issues`,
    navigator: window.navigator,
  }),
  computed: {
    ...mapState([`session`]),
    ...mapGetters([`networkSlug`]),
    ...mapGetters({ networkId: `network` }),
    submitCaption() {
      return {
        connect: "Sign In",
        detect: "Waiting for Ledger",
      }[this.status]
    },
    isWindows() {
      return this.navigator.platform.indexOf("Win") > -1
    },
    isLinux() {
      return this.navigator.platform.indexOf("Lin") > -1
    },
    hasHIDEnabled() {
      return !!this.navigator.hid
    },
    isChrome() {
      const ua = navigator.userAgent.toLowerCase()
      return /chrome|crios/.test(ua) && !/edge|opr\//.test(ua)
    },
  },
  methods: {
    async signIn() {
      this.connectionError = null
      this.status = `detect`
      this.address = null
      try {
        this.address = await getAddressFromLedger(this.networkId, this.$store)
        this.$router.push({
          name: "portfolio",
          params: {
            networkId: this.networkSlug,
          },
        })
      } catch (error) {
        this.status = `connect`
        this.connectionError = error.message
        Sentry.captureException(error)
        return
      }

      await this.$store.dispatch(`signIn`, {
        sessionType: `ledger`,
        address: this.address,
        networkId: this.networkId,
      })
    },
    onCopy() {
      this.copySuccess = true
      setTimeout(() => {
        this.copySuccess = false
      }, 2500)
    },
  },
}
</script>
<style scoped>
.error-message {
  color: var(--danger);
  font-size: var(--sm);
  font-style: italic;
  margin-bottom: 0;
  padding-top: 1rem;
}

.install-notes {
  flex-direction: column;
}

.ledger-install {
  font-size: var(--sm);
  margin-bottom: 0;
}

.address {
  color: var(--link);
  font-weight: 500;
  font-size: 14px;
  white-space: nowrap;
}

.form-message {
  font-size: var(--sm);
  font-weight: 500;
  font-style: italic;
  color: var(--dim);
  display: inline-block;
}

.form-message.notice {
  border-radius: 0.25rem;
  border: 1px solid var(--bc-dim);
  background-color: #1c223e;
  font-weight: 400;
  margin: 2rem 0;
  padding: 1rem 1rem;
  font-size: 14px;
  font-style: normal;
  width: 100%;
}

.copy-feature-link {
  display: initial;
  font-size: 0.8rem;
  cursor: pointer;
  margin-bottom: 0.2rem;
  color: var(--link);
}

.copy-feature-link .material-icons {
  font-size: 12px;
}

.copy-feature-link .copied {
  padding-bottom: 2px;
  padding-right: 0;
  transition: opacity 500ms ease;
  color: var(--success);
  opacity: 0;
}

.copy-feature-link .copied.active {
  opacity: 1;
}

.session-main .button {
  margin: 2rem auto 0;
}
</style>
