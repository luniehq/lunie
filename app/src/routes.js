/**
 * In this module we took care of the definition of our routes, with parameters, children and component related to them
 * @module routes
 */

/**
 * Routes are all defined here
 */
/* istanbul ignore file */
export default (store) => {
  return [
    {
      path: `/`,
      beforeEnter: async (to, from, next) => {
        const userSignedIn = await store.dispatch("checkSession")
        if (userSignedIn) {
          next("/notifications")
        } else {
          next("/cosmos-hub/portfolio")
        }
      },
    },
    {
      path: `/networks`,
      name: `networks`,
      component: () => import(`./components/network/PageNetworks`),
      meta: {
        requiresAuth: false,
      },
    },
    {
      path: `/notification-wall`,
      name: `notification-wall`,
      components: {
        session: () => import(`./components/account/Paywall`),
      },
    },
    {
      path: `/notifications`,
      name: `notifications`,
      components: {
        session: () => import(`./components/notifications/PageNotifications`),
      },
      meta: {
        feature: "notifications",
      },
    },
    {
      path: `/login`,
      name: `login`,
      components: {
        session: () => import(`./components/session/TmSessionSignIn`),
      },
      meta: {
        feature: "session",
      },
    },
    {
      path: `/select-network`,
      name: `select-network`,
      components: {
        session: () => import(`./components/common/TmSelectNetwork`),
      },
      meta: {
        feature: "session",
      },
    },
    {
      path: `/select-network/recover`,
      name: `select-network-recover`,
      components: {
        session: () => import(`./components/common/TmSelectNetwork`),
      },
      meta: {
        feature: "session",
      },
    },
    {
      path: `/select-network/create`,
      name: `select-network-create`,
      components: {
        session: () => import(`./components/common/TmSelectNetwork`),
      },
      meta: {
        feature: "session",
      },
    },
    {
      path: `/create`,
      name: `create`,
      components: {
        session: () => import(`./components/session/TmSessionSignUp`),
      },
      meta: {
        feature: "session",
      },
    },
    {
      path: `/create/password`,
      name: `create-password`,
      components: {
        session: () => import(`./components/session/TmSessionSignUpPassword`),
      },
      meta: {
        feature: "session",
      },
      beforeEnter: (to, from, next) => {
        if (from.name === `create`) {
          next()
        } else {
          next({ path: `/create` })
        }
      },
    },
    {
      path: `/create/confirm`,
      name: `create-confirm`,
      components: {
        session: () => import(`./components/session/TmSessionSignUpSeed`),
      },
      meta: {
        feature: "session",
      },
      beforeEnter: (to, from, next) => {
        if (from.name === `create-password`) {
          next()
        } else {
          next({ path: `/create` })
        }
      },
    },
    {
      path: `/recover`,
      name: `recover`,
      components: {
        session: () => import(`./components/session/TmSessionImport`),
      },
      meta: {
        feature: "session",
      },
    },
    {
      path: `/recover/name`,
      name: `recover-name`,
      components: {
        session: () => import(`./components/session/TmSessionImportName`),
      },
      meta: {
        feature: "session",
      },
      beforeEnter: (to, from, next) => {
        if (from.name === `recover`) {
          next()
        } else {
          next({ path: `/recover` })
        }
      },
    },
    {
      path: `/recover/password`,
      name: `recover-password`,
      components: {
        session: () => import(`./components/session/TmSessionImportPassword`),
      },
      meta: {
        feature: "session",
      },
      beforeEnter: (to, from, next) => {
        if (from.name === `recover-name`) {
          next()
        } else {
          next({ path: `/recover` })
        }
      },
    },
    {
      path: `/explore`,
      name: `explore`,
      components: {
        session: () => import(`./components/session/TmSessionExplore`),
      },
      meta: {
        feature: "session",
      },
    },
    {
      path: `/ledger`,
      name: `ledger`,
      components: {
        session: () => import(`./components/session/TmSessionHardware`),
      },
      meta: {
        feature: "session",
      },
    },
    {
      path: `/extension/:address/:network`,
      name: `extension-signin`,
      beforeEnter: function (to, from, next) {
        /* istanbul ignore next */
        return extensionSignIn({ to, from, next }, store)
      },
    },
    {
      path: `/extension`,
      name: `extension`,
      components: {
        session: () => import(`./components/session/TmSessionExtension`),
      },
      meta: {
        feature: "session",
      },
    },
    { path: `/404`, component: () => import(`./components/common/Page404`) },
    {
      path: `/privacy`,
      beforeEnter() {
        location.href = `https://lunie.io/privacy`
      },
    },
    {
      path: `/terms`,
      beforeEnter() {
        location.href = `https://lunie.io/terms`
      },
    },
    {
      path: `/security`,
      beforeEnter() {
        location.href = `https://lunie.io/security`
      },
    },
    {
      path: `/about`,
      beforeEnter() {
        location.href = `https://lunie.io/about`
      },
    },
    {
      path: `/careers`,
      beforeEnter() {
        location.href = `https://angel.co/company/lunie`
      },
    },
    {
      path: `/feature-not-available/:feature`,
      component: () => import(`./components/common/PageFeatureNotAvailable`),
      props: true,
      meta: {
        networkSpecificRoute: true,
      },
    },
    {
      path: `/feature-not-present/:feature`,
      component: () => import(`./components/common/FeatureNotPresent`),
    },
    {
      path: `/sign-up-email`,
      name: `sign-in-modal`,
      component: require("account/SignInModal").default,
    },
    {
      path: `/magic-link`,
      name: `magic-link-sent`,
      component: require("account/MagicLinkSentModal").default,
    },
    {
      path: `/reveal/:address`,
      name: `reveal`,
      component: require("account/RevealSeedModal").default,
    },
    {
      path: `/delete/:address/:addressNetworkId`,
      name: `delete`,
      component: require("account/ForgetAccountModal").default,
    },
    {
      path: `/email-authentication`,
      name: `email-authentication`,
      component: () => import(`./components/account/EmailAuthentication`),
    },
    {
      path: `/welcome`,
      name: `welcome`,
      component: () => import(`./components/common/CardSignInRequired`),
      meta: {
        networkSpecificRoute: true,
      },
    },
    {
      path: `/manage-accounts`,
      name: `manage-accounts-modal`,
      component: () => import(`./components/account/ManageAccountsModal`),
      meta: {
        networkSpecificRoute: true,
      },
    },
    {
      path: `/:networkId`,
      component: () => import(`./components/common/NetworkSetter`),
      redirect: `/:networkId/portfolio`,
      children: [
        {
          path: `proposals`,
          name: `proposals`,
          meta: {
            feature: "proposals",
            networkSpecificRoute: true,
          },
          component: () => import(`./components/governance/PageProposals`),
        },
        // for deprecated routes
        {
          path: `governance/proposals`,
          redirect: `/proposals`,
        },
        {
          path: `proposals/:proposalId`,
          name: `proposal`,
          meta: {
            feature: "proposals",
            networkSpecificRoute: true,
          },
          component: () => import(`./components/governance/PageProposal`),
        },
        // for deprecated routes
        {
          path: `governance/proposals/:proposalId`,
          redirect: `/proposals/:proposalId`,
        },
        {
          path: `validators`,
          name: `validators`,
          meta: {
            feature: "validators",
            networkSpecificRoute: true,
          },
          component: () => import(`./components/staking/PageValidators`),
        },
        // for deprecated routes
        {
          path: `staking/validators`,
          redirect: `/validators`,
        },
        {
          path: `validators/:validator`,
          name: `validator`,
          meta: {
            feature: "validators",
            networkSpecificRoute: true,
          },
          component: () => import(`./components/staking/PageValidator`),
        },
        {
          path: `staking/validators/:validator`,
          redirect: `/validators/:validator`,
        },
        {
          path: `portfolio`,
          name: `portfolio`,
          component: () => import(`./components/common/PagePortfolio`),
          meta: {
            requiresAuth: true,
            feature: "portfolio",
            networkSpecificRoute: true,
          },
        },
        {
          path: `transactions`,
          name: `transactions`,
          component: () => import(`./components/transactions/PageTransactions`),
          meta: {
            requiresAuth: true,
            feature: "activity",
            networkSpecificRoute: true,
          },
        },
        {
          path: `blocks/:height`,
          name: `block`,
          component: () => import(`./components/network/PageBlock`),
          meta: {
            feature: "blocks",
            networkSpecificRoute: true,
          },
        },
        { path: `*`, component: () => import(`./components/common/Page404`) },
      ],
    },
  ]
}

// handle direct sign in from the extension via deeplink
export async function extensionSignIn({ to, next }, store) {
  let network = store.getters.networks.find(
    ({ id }) => id === to.params.network
  )

  await store.dispatch(`signIn`, {
    sessionType: `extension`,
    address: to.params.address,
    networkId: to.params.network,
  })
  next(`/${network.slug}/portfolio`)
}
