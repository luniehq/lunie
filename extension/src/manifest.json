{
  "name": "Lunie Browser Extension",
  "description": "A browser extension for Lunie.io",
  "version": null,
  "manifest_version": 2,
  "icons": {
    "128": "icons/128x128.png"
  },
  "browser_action": {
    "default_title": "Lunie Browser Extension",
    "default_popup": "popup/popup.html"
  },
  "background": {
    "scripts": [
      "background.js"
    ]
  },
  "sandbox": {
    "pages": [
      "popup/sandbox.html"
    ]
  },
  "web_accessible_resources": [
    "fonts/*.woff2"
  ],
  "content_security_policy": "script-src 'self' 'unsafe-eval' https://www.google-analytics.com/analytics.js; object-src 'self'; connect-src 'self' wss://production-api.lunie.io/graphql https://production-api.lunie.io wss://staging-api.lunie.io/graphql https://staging-api.lunie.io https://monitoring.lunie.io:9000 wss://kusama-rpc.polkadot.io wss://rpc.polkadot.io wss://westend-rpc.polkadot.io wss://westend.polkastats.io/rpc https://www.google-analytics.com https://static.intercomassets.com https://js.intercomcdn.com/images;",
  "content_scripts": [
    {
      "matches": [
        "http://localhost/*",
        "https://localhost/*",
        "https://lunie.io/*",
        "https://app.lunie.io/*",
        "https://wallet.e-money.com/*",
        "https://lunie.cosmos.network/*",
        "https://lunie-akash.vitwit.com/*",
        "https://www.lunie.io/*",
        "https://*.netlify.com/*",
        "https://*.netlify.app/*"
      ],
      "js": [
        "contentScript.js"
      ]
    }
  ],
  "permissions": [
    "tabs"
  ]
}