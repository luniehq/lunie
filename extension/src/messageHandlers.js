const {
  getWalletIndex,
  getStoredWallet,
  testPassword,
  removeWallet,
  removeFromStorage
} = require('@lunie/cosmos-keys')
const {
  default: TransactionManager
} = require('app/src/signing/transaction-manager')
const { getPolkadotAPI } = require('../../common/polkadotApiConnector')

export async function signMessageHandler(
  signRequestQueue,
  event,
  sender,
  sendResponse
) {
  switch (event.type) {
    case 'LUNIE_SIGN_REQUEST_CANCEL': {
      signRequestQueue.unqueueSignRequestForTab(sender.tab.id)
      break
    }
    case 'LUNIE_GET_SIGN_QUEUE': {
      sendAsyncResponseToLunie(sender.tab.id, {
        type: 'LUNIE_GET_SIGN_QUEUE_RESPONSE',
        payload: {
          amount: signRequestQueue.getQueueLength()
        }
      })
      break
    }
    case 'LUNIE_SIGN_REQUEST': {
      const {
        // old messaging
        signMessage, //DEPRECATE
        displayedProperties, //DEPRECATE

        messageType,
        message,
        transactionData,
        senderAddress,
        network
      } = event.payload
      const wallet = getWalletFromIndex(getWalletIndex(), senderAddress)
      if (!wallet) {
        throw new Error('No wallet found matching the sender address.')
      }

      signRequestQueue.queueSignRequest({
        signMessage, // DEPRECATE
        displayedProperties, //DEPRECATE

        messageType,
        message,
        transactionData,
        senderAddress,
        network,
        tabID: sender.tab.id
      })
      break
    }
    case 'SIGN': {
      const {
        messageType,
        message,
        transactionData,
        senderAddress,
        network,
        password,
        id,
        HDPath,
        curve
      } = event.payload

      const transactionManager = new TransactionManager()
      let polkadotAPI
      if (network.network_type === `polkadot`) {
        polkadotAPI = await getPolkadotAPI(network)
      }
      try {
        const broadcastableObject = await transactionManager.createAndSignLocally(
          messageType,
          message,
          transactionData,
          senderAddress,
          network,
          'local',
          password,
          polkadotAPI,
          HDPath,
          curve
        )
        const { tabID } = signRequestQueue.unqueueSignRequest(id)
        sendAsyncResponseToLunie(tabID, {
          type: 'LUNIE_SIGN_REQUEST_RESPONSE',
          payload: broadcastableObject
        })

        sendResponse() // to popup
      } catch (error) {
        sendResponse({ error: error.message }) // to popup
      }

      break
    }
    case 'GET_SIGN_REQUEST': {
      sendResponse(signRequestQueue.getSignRequest())
      break
    }
    case 'REJECT_SIGN_REQUEST': {
      const { id, tabID } = event.payload
      sendAsyncResponseToLunie(tabID, {
        type: 'LUNIE_SIGN_REQUEST_RESPONSE',
        payload: { rejected: true }
      })
      signRequestQueue.unqueueSignRequest(id)
      sendResponse() // to popup
      break
    }
  }
}
export async function walletMessageHandler(message, sender, sendResponse) {
  switch (message.type) {
    case 'GET_WALLETS': {
      sendResponse(getWalletIndex(true))
      break
    }
    case 'DELETE_WALLET': {
      const { address, password } = message.payload
      removeWallet(address, password)
      sendResponse()
      break
    }
    case 'DELETE_WALLET_WITHOUT_PASSWORD': {
      const { address } = message.payload
      removeFromStorage(address)
      sendResponse()
      break
    }
    case 'TEST_PASSWORD': {
      const { address, password } = message.payload
      try {
        testPassword(address, password)
        sendResponse(true)
      } catch (error) {
        sendResponse(false)
      }
      break
    }
    case 'GET_WALLET': {
      const { address, password } = message.payload
      try {
        const wallet = await getStoredWallet(address, password)
        sendResponse(wallet)
      } catch (error) {
        sendResponse()
      }
      break
    }
  }
}

// for responses that take some time like for a sign request we can't use simple responses
// we instead send a messsage to the sending tab
function sendAsyncResponseToLunie(tabId, { type, payload }) {
  chrome.tabs.sendMessage(tabId, { type, payload })
}

function getWalletFromIndex(walletIndex, address) {
  return walletIndex.find(
    ({ address: storedAddress }) => storedAddress === address
  )
}
