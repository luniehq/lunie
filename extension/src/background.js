import 'core-js/stable'
import 'regenerator-runtime/runtime'
import { signMessageHandler, walletMessageHandler } from './messageHandlers'
import SignRequestQueue from './requests'
import { bindRequestsToTabs } from './tabsHandler'

global.browser = require('webextension-polyfill')

const extensionHost = location.origin
const whitelisted = [
  'https://app.lunie.io',
  'https://wallet.e-money.com',
  'https://lunie.cosmos.network',
  'https://lunie-akash.vitwit.com',
  /https:\/\/\w+--lunieio.netlify.com/, // to use the extension with deployment previews
  extensionHost
]
if (process.env.NODE_ENV === 'development') {
  whitelisted.push('https://localhost')
  whitelisted.push('http://localhost')
}

const signRequestQueue = new SignRequestQueue()
signRequestQueue.unqueueSignRequest('')

// main message handler
chrome.runtime.onMessage.addListener((message, sender, sendResponse) => {
  if (!senderAllowed(sender)) {
    console.error('Sender is not whitelisted')
    return
  }

  try {
    signMessageHandler(signRequestQueue, message, sender, sendResponse)
    walletMessageHandler(message, sender, sendResponse)
  } catch (error) {
    // Return this as rejected
    console.error('Error with request', error)
    sendResponse({ error: error.message })
  }

  return true
})

// only allow whitelisted websites to send us messages
function senderAllowed(sender) {
  // if sender.tab is not defined, the message comes from the extension
  if (sender.tab && !whitelistedChecker(sender.tab.url)) {
    return false
  }
  return true
}

const whitelistedChecker = (url) => {
  return !!whitelisted.find((whitelistedUrl) => {
    // check regexps
    if (whitelistedUrl instanceof RegExp) {
      whitelistedUrl.exec(url)
      return whitelistedUrl.lastIndex === 0 // check if the regex matches on index 0 (to avoid any possible hack later on)
    }
    // prefer normal strings as easier to read and error check
    return url.startsWith(whitelistedUrl)
  })
}

bindRequestsToTabs(signRequestQueue, whitelistedChecker)
