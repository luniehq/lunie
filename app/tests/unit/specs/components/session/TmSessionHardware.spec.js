import Vuex from "vuex"
import Vuelidate from "vuelidate"
import { mount, createLocalVue } from "@vue/test-utils"
import TmSessionHardware from "session/TmSessionHardware"
import { focusParentLast } from "src/directives"

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(Vuelidate)
localVue.directive(`tooltip`, () => {})
localVue.directive(`focus`, () => {})
localVue.directive("focus-last", focusParentLast)
localVue.directive("clipboard", () => {})

jest.mock("scripts/ledger", () => ({
  getAddressFromLedger: () => "cosmos1234",
}))

describe(`TmSessionHardware`, () => {
  let wrapper, store

  beforeEach(() => {
    store = new Vuex.Store({
      state: {
        session: { browserWithLedgerSupport: true },
      },
    })

    wrapper = mount(TmSessionHardware, {
      localVue,
      store,
      mocks: {
        router: {
          push: jest.fn(),
        },
      },
      stubs: [`router-link`],
    })
    store.commit = jest.fn()
  })

  describe(`shows the ledger conection view`, () => {
    it(`with no errors`, () => {
      expect(wrapper.element).toMatchSnapshot()
    })

    it(`when there're errors`, async () => {
      wrapper.vm.connectionError = `No Ledger found`
      await wrapper.vm.$nextTick()
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  it(`should show a state indicator for different states of the hardware connection`, () => {
    wrapper.setData({ status: `connect` })
    expect(wrapper.html()).toMatchSnapshot()

    wrapper.setData({ status: `detect` })
    expect(wrapper.html()).toMatchSnapshot()

    wrapper.setData({ status: `confirmAddress` })
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe(`sign in`, () => {
    it(`signs in if Ledger is connected and app is open and address is confirmed`, async () => {
      const $store = {
        dispatch: jest.fn(() => "cosmos1234"),
      }
      const $router = {
        push: jest.fn(),
      }
      const self = {
        $store,
        $router,
        status: `connect`,
        connectionError: null,
        setStatus: jest.fn(),
        setConnectionError: jest.fn((error) => (self.connectionError = error)),
        confirmAddress: jest.fn(() => true),
      }
      await TmSessionHardware.methods.signIn.call(self)
      expect(self.connectionError).toBeNull()
      expect(self.$store.dispatch).toHaveBeenCalledWith(`signIn`, {
        sessionType: `ledger`,
        address: "cosmos1234",
      })
    })

    it(`doesn't sign in if ledger not connected`, async () => {
      jest.resetModules()
      jest.doMock("scripts/ledger", () => ({
        getAddressFromLedger: () =>
          Promise.reject(new Error(`No Ledger found`)),
      }))
      const TmSessionHardware = require("session/TmSessionHardware").default

      const $store = {
        dispatch: jest.fn(() => "cosmos1234"),
      }

      const self = {
        $store,
        status: `connect`,
        connectionError: null,
        setStatus: jest.fn(),
        setConnectionError: jest.fn((error) => (self.connectionError = error)),
      }
      await TmSessionHardware.methods.signIn.call(self)
      expect(self.connectionError).toBe(`No Ledger found`)
      expect(self.$store.dispatch).not.toHaveBeenCalledWith(
        `signIn`,
        expect.objectContaining({})
      )
    })

    it(`does show the instructions to enable HID on Windows`, () => {
      wrapper.setData({
        navigator: {
          hid: undefined,
          platform: "Win64",
          userAgent: "Chrome",
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
      expect(wrapper.html()).toContain(
        "Due to recent Ledger updates, using a Ledger on Windows"
      )
    })

    it(`does show the instructions to fix connection issues for Linux users`, () => {
      wrapper.setData({
        navigator: {
          platform: "Linux i686",
          userAgent: "Chrome",
        },
      })

      expect(wrapper.html()).toMatchSnapshot()
      expect(wrapper.html()).toContain(
        "Since we switched to WebUSB Linux users may experience connection"
      )
    })

    describe(`onCopy`, () => {
      it(`should set and reset copySuccess`, () => {
        jest.useFakeTimers()
        wrapper.vm.onCopy() // old test style to make timer work
        expect(wrapper.vm.copySuccess).toBe(true)

        jest.runAllTimers()
        expect(wrapper.vm.copySuccess).toBe(false)
      })
    })
  })
})
