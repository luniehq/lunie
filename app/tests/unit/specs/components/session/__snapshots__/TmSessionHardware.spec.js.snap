// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TmSessionHardware should show a state indicator for different states of the hardware connection 1`] = `
<div tabindex="0" class="session-frame" name="component-fade" mode="out-in">
  <div class="session-outer-container"><i class="material-icons notranslate circle modal-icon">vpn_key</i>
    <div class="session">
      <div class="session-header"><a class=""><i class="material-icons notranslate circle back">arrow_back</i></a>
        <div class="session-close"><a><i class="material-icons notranslate circle back">close</i></a></div>
      </div>
      <h2 class="session-title">Use my Ledger Nano</h2>
      <div class="session-main">
        <div class="tm-hardware-state">
          <div class="tm-hardware-state-content">
            <!---->
            <div class="tm-hardware-state__label">
              <p>
                Please plug in your Ledger&nbsp;Nano and open the Cosmos Ledger
                app
              </p>
              <!----> <button class="button">
                Sign In
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TmSessionHardware should show a state indicator for different states of the hardware connection 2`] = `
<div tabindex="0" class="session-frame" name="component-fade" mode="out-in">
  <div class="session-outer-container"><i class="material-icons notranslate circle modal-icon">vpn_key</i>
    <div class="session">
      <div class="session-header"><a class=""><i class="material-icons notranslate circle back">arrow_back</i></a>
        <div class="session-close"><a><i class="material-icons notranslate circle back">close</i></a></div>
      </div>
      <h2 class="session-title">Use my Ledger Nano</h2>
      <div class="session-main">
        <div class="tm-hardware-state">
          <div class="tm-hardware-state-content"><img src="~assets/images/loader.svg" alt="a small spinning circle to display loading" class="tm-hardware-state__icon">
            <div class="tm-hardware-state__label">
              <p>
                Please plug in your Ledger&nbsp;Nano and open the Cosmos Ledger
                app
              </p>
              <!----> <button class="button" disabled="disabled">
                Waiting for Ledger
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TmSessionHardware should show a state indicator for different states of the hardware connection 3`] = `
<div tabindex="0" class="session-frame" name="component-fade" mode="out-in">
  <div class="session-outer-container"><i class="material-icons notranslate circle modal-icon">vpn_key</i>
    <div class="session">
      <div class="session-header"><a class=""><i class="material-icons notranslate circle back">arrow_back</i></a>
        <div class="session-close"><a><i class="material-icons notranslate circle back">close</i></a></div>
      </div>
      <h2 class="session-title">Use my Ledger Nano</h2>
      <div class="session-main">
        <div class="tm-hardware-state">
          <div class="tm-hardware-state-content"><img src="~assets/images/loader.svg" alt="a small spinning circle to display loading" class="tm-hardware-state__icon">
            <div class="tm-hardware-state__label">
              <!---->
              <!----> <button class="button" disabled="disabled">

              </button></div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TmSessionHardware shows the ledger conection view when there're errors 1`] = `
<div
  class="session-frame"
  mode="out-in"
  name="component-fade"
  tabindex="0"
>
  <div
    class="session-outer-container"
  >
    <i
      class="material-icons notranslate circle modal-icon"
    >
      vpn_key
    </i>
     
    <div
      class="session"
    >
      <div
        class="session-header"
      >
        <a
          class=""
        >
          <i
            class="material-icons notranslate circle back"
          >
            arrow_back
          </i>
        </a>
         
        <div
          class="session-close"
        >
          <a>
            <i
              class="material-icons notranslate circle back"
            >
              close
            </i>
          </a>
        </div>
      </div>
       
      <h2
        class="session-title"
      >
        Use my Ledger Nano
      </h2>
       
      <div
        class="session-main"
      >
        <div
          class="tm-hardware-state"
        >
          <div
            class="tm-hardware-state-content"
          >
            <!---->
             
            <div
              class="tm-hardware-state__label"
            >
              <p>
                
            Please plug in your Ledger Nano and open the Cosmos Ledger
            app
          
              </p>
               
              <p
                class="error-message"
              >
                
          No Ledger found
        
              </p>
               
              <button
                class="button"
              >
                
    Sign In
  
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TmSessionHardware shows the ledger conection view with no errors 1`] = `
<div
  class="session-frame"
  mode="out-in"
  name="component-fade"
  tabindex="0"
>
  <div
    class="session-outer-container"
  >
    <i
      class="material-icons notranslate circle modal-icon"
    >
      vpn_key
    </i>
     
    <div
      class="session"
    >
      <div
        class="session-header"
      >
        <a
          class=""
        >
          <i
            class="material-icons notranslate circle back"
          >
            arrow_back
          </i>
        </a>
         
        <div
          class="session-close"
        >
          <a>
            <i
              class="material-icons notranslate circle back"
            >
              close
            </i>
          </a>
        </div>
      </div>
       
      <h2
        class="session-title"
      >
        Use my Ledger Nano
      </h2>
       
      <div
        class="session-main"
      >
        <div
          class="tm-hardware-state"
        >
          <div
            class="tm-hardware-state-content"
          >
            <!---->
             
            <div
              class="tm-hardware-state__label"
            >
              <p>
                
            Please plug in your Ledger Nano and open the Cosmos Ledger
            app
          
              </p>
               
              <!---->
               
              <button
                class="button"
              >
                
    Sign In
  
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TmSessionHardware sign in does show the instructions to enable HID on Windows 1`] = `
<div tabindex="0" class="session-frame" name="component-fade" mode="out-in">
  <div class="session-outer-container"><i class="material-icons notranslate circle modal-icon">vpn_key</i>
    <div class="session">
      <div class="session-header"><a class=""><i class="material-icons notranslate circle back">arrow_back</i></a>
        <div class="session-close"><a><i class="material-icons notranslate circle back">close</i></a></div>
      </div>
      <h2 class="session-title">Use my Ledger Nano</h2>
      <div class="session-main">
        <div class="tm-hardware-state">
          <div class="tm-hardware-state-content">
            <!---->
            <div class="tm-hardware-state__label">
              Due to recent Ledger updates, using a Ledger on Windows now requires
              "Experimental Web Platform features" to be enabled.
              <!---->
              <!----> <button class="button">
                Sign In
              </button></div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`TmSessionHardware sign in does show the instructions to fix connection issues for Linux users 1`] = `
<div tabindex="0" class="session-frame" name="component-fade" mode="out-in">
  <div class="session-outer-container"><i class="material-icons notranslate circle modal-icon">vpn_key</i>
    <div class="session">
      <div class="session-header"><a class=""><i class="material-icons notranslate circle back">arrow_back</i></a>
        <div class="session-close"><a><i class="material-icons notranslate circle back">close</i></a></div>
      </div>
      <h2 class="session-title">Use my Ledger Nano</h2>
      <div class="session-main">
        <div class="tm-hardware-state">
          <div class="tm-hardware-state-content">
            <!---->
            <div class="tm-hardware-state__label">
              Since we switched to WebUSB Linux users may experience connection
              issues with their devices.
              <br> <br>
              Please visit the following site to learn more about how to fix them:
              <div class="copy-feature-link">
                https://support.ledger.com/hc/en-us/articles/360019301813-Fix-USB-issues
                <i class="material-icons notranslate copied">
                  check
                </i></div> <br> <br>
              <!----> <button class="button">
                Sign In
              </button></div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
